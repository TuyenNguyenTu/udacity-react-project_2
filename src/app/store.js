import { configureStore } from '@reduxjs/toolkit';
import authedSlice from '../features/authSlice'
import questionSlice from '../features/questionSlice'
import userSlice from '../features/userSlice'
import loadingSlice from '../features/loadingSlice'

export const store = configureStore({
  reducer: {
    auth: authedSlice,
    question: questionSlice,
    user: userSlice,
    loading: loadingSlice
  },
});
