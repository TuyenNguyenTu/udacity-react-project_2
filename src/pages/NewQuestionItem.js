import React, { useState } from "react";
import * as Datas from "../Data";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import * as questionsAction from "../store/actions/questionsAction";
import * as userAction from "../store/actions/userAction";
const selectAuth = (state) => state.auth;
const selectQuesions = (state) => state.questions;
const selectUsers = (state) => state.users;

export default function NewQuestionItem() {
  const [optionOneText, setOptionOneText] = useState();
  const [optionTwoText, setOptionTwoText] = useState();
  const auth = useSelector(selectAuth);
  const questions = useSelector(selectQuesions);
  const users = useSelector(selectUsers);
  const [message, setMessage] = useState("");
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const handleSubmit = (e) => {
    if (!optionOneText || !optionTwoText) {
      setMessage("please input optionText");
      return;
    }
    Datas._saveQuestion({
      optionOneText: optionOneText,
      optionTwoText: optionTwoText,
      author: auth.user.id,
    })
      .then((res) => {
        if (res) {
          const newQuestions = { ...questions, [res.id]: res };
          dispatch(questionsAction.setupQuestions(newQuestions));
          const newusers = {
            ...users,
            [res.author]: {
              ...users[res.author],
              questions: [...users[res.author].questions, res.id],
            },
          };
          dispatch(userAction.setUsers(newusers));
          navigate("/home");
        }
      })
      .catch((error) => {
        console.log(error);
        setMessage("can not save question");
      });
  };

  return (
    <div className="container">
      <div className="py-5 text-center">
        <h2>Would You Rather</h2>
        <p className="lead">Create Your Own Poll</p>
        <div className="row text-center w-100">
          <div className="col-12">
            <span className="">{message}</span>
          </div>
          <div className="col-12 row mb-3">
            <div className="col-2"><label className="form-label">First Option:</label></div>
            <div className="col-9">
              <input
              type="text"
              className="form-control"
              placeholder="Option One"
              data-testid="OptionOne"
              onChange={(event) => setOptionOneText(event.target.value)}
            />
            </div>
          </div>
          <div className="col-12 row mb-3">
            <div className="col-2"><label className="form-label">Second Option:</label></div>
            <div className="col-9">
              <input
                type="text"
                className="form-control"
                placeholder="Option Two"
                data-testid="OptionTwo"
                onChange={(event) => setOptionTwoText(event.target.value)}
              />
            </div>
          </div>
          <div className="col-12 text-center">
            <button
              className="btn btn-light"
              type="button"
              onClick={handleSubmit}
            >
              Submit
            </button>
          </div>          
        </div>
      </div>
    </div>
  );
}
