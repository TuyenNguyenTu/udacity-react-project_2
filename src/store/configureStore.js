import logger from "redux-logger";
import rootReducer from "./reducers";
import {  configureStore } from "@reduxjs/toolkit";


let store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({ serializableCheck: false }).concat(logger)
});

export default store; 
