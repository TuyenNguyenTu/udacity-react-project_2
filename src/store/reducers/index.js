import { combineReducers } from "@reduxjs/toolkit";
import authenticateReducer from "./authenticateReducer";
import questionReducer from "./questionsReducer";
import usersReducer from "./usersReducer";

const rootReducer = combineReducers({
  auth: authenticateReducer,
  questions: questionReducer,
  users: usersReducer,
});

export default rootReducer;
