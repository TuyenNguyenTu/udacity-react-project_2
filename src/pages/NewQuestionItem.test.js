import { fireEvent, render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { store } from "../store";
import NewQuestionItem from "./NewQuestionItem";

describe("NewQuestionItem Test", () => {
  test("should render the component", () => {
    const view = render(
      <BrowserRouter>
        <Provider store={store}>
          <NewQuestionItem />
        </Provider>
      </BrowserRouter>
    );
    expect(view).toBeDefined();
    expect(view).toMatchSnapshot();
  });

  test("should render correct data", () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <NewQuestionItem />
        </Provider>
      </BrowserRouter>
    );

    const optionOneInputElement = screen.getByTestId("OptionOne");
    const optionTwoInputElement = screen.getByTestId("OptionTwo");

    fireEvent.change(optionOneInputElement, { target: { value: "test 1" } });
    fireEvent.change(optionTwoInputElement, {
      target: { value: "test 2" },
    });
    expect(optionOneInputElement.value).toBe("test 1");
    expect(optionTwoInputElement.value).toBe("test 2");
  });
});
