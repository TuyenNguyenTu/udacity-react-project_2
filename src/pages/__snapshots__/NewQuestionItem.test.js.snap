// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewQuestionItem Test should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="py-5 text-center"
        >
          <h2>
            Would You Rather
          </h2>
          <p
            class="lead"
          >
            Create Your Own Poll
          </p>
          <div
            class="row text-center w-100"
          >
            <div
              class="col-12"
            >
              <span
                class=""
              />
            </div>
            <div
              class="col-12 row mb-3"
            >
              <div
                class="col-2"
              >
                <label
                  class="form-label"
                >
                  First Option:
                </label>
              </div>
              <div
                class="col-9"
              >
                <input
                  class="form-control"
                  data-testid="OptionOne"
                  placeholder="Option One"
                  type="text"
                />
              </div>
            </div>
            <div
              class="col-12 row mb-3"
            >
              <div
                class="col-2"
              >
                <label
                  class="form-label"
                >
                  Second Option:
                </label>
              </div>
              <div
                class="col-9"
              >
                <input
                  class="form-control"
                  data-testid="OptionTwo"
                  placeholder="Option Two"
                  type="text"
                />
              </div>
            </div>
            <div
              class="col-12 text-center"
            >
              <button
                class="btn btn-light"
                type="button"
              >
                Submit
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="py-5 text-center"
      >
        <h2>
          Would You Rather
        </h2>
        <p
          class="lead"
        >
          Create Your Own Poll
        </p>
        <div
          class="row text-center w-100"
        >
          <div
            class="col-12"
          >
            <span
              class=""
            />
          </div>
          <div
            class="col-12 row mb-3"
          >
            <div
              class="col-2"
            >
              <label
                class="form-label"
              >
                First Option:
              </label>
            </div>
            <div
              class="col-9"
            >
              <input
                class="form-control"
                data-testid="OptionOne"
                placeholder="Option One"
                type="text"
              />
            </div>
          </div>
          <div
            class="col-12 row mb-3"
          >
            <div
              class="col-2"
            >
              <label
                class="form-label"
              >
                Second Option:
              </label>
            </div>
            <div
              class="col-9"
            >
              <input
                class="form-control"
                data-testid="OptionTwo"
                placeholder="Option Two"
                type="text"
              />
            </div>
          </div>
          <div
            class="col-12 text-center"
          >
            <button
              class="btn btn-light"
              type="button"
            >
              Submit
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
